File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.swift
import Foundation

func solution(_ numbers:[Int], _ target:Int) -> Int {
    dfs(numbers, target, 0, 0)
}

func dfs(_ numbers:[Int], _ target:Int, _ index: Int, _ sum: Int) -> Int {
    if index == numbers.count {
        if sum == target {
            return 1
        }
        return 0
    }

    let left = dfs(numbers, target, index + 1, sum + numbers[index])
    let right = dfs(numbers, target, index + 1, sum - numbers[index])
    
    return left + right
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    answer = 0
    max_depth = len(numbers)

    def DFS(depth, total_sum):
        nonlocal answer
        if max_depth == depth:
            if target == total_sum:
                answer += 1
            return

        DFS(depth + 1, total_sum + numbers[depth])
        DFS(depth + 1, total_sum - numbers[depth])


    DFS(0, 0)
    return answer
